### PR TITLE
ENH: `stats.make_distribution`: generate a `ContinuousDistribution` from existing `rv_continuous`

### PR DESCRIPTION
**Pull request by mdhaber**
#### What does this implement/fix?
This adds a function that accepts an instance of `stats.rv_continuous` and returns a subclass of `stats.ContinuousDistribution`. The `ContinuousDistribution` uses overridden methods of the `rv_continuous` class where possible (e.g. `pdf`, `logpdf`) but uses its own generic method implementations (which are often faster and more accurate).

As an example inspired by gh-18799:
```python3
from scipy import stats
GH = stats.make_distribution(stats.gausshyper)
X = stats.gausshyper(a=1.5, b=2.5, c=2, z=0)
Y = GH(a=1.5, b=2.5, c=2, z=0)
```
`X` is a "frozen" distribution object representing the Gauss hypergeometric distribution and `Y` is a corresponding instance of a `ContinuousDistribution`. Since `gausshyper_gen` defines `_pdf`, the `pdf` methods of the two objects behave identically:
```python3
X.pdf(0.3), Y.pdf(0.3)
# (np.float64(1.6337206456367985), np.float64(1.6337206456367985))
```
but because `gausshyper_gen` does not define `_cdf`, `Y` will use the `ContinuousDistribution` framework to evaluate the CDF, which tends to be much faster and more accurate than the generic implementation in the `rv_continuous` framework.
<details>

```python3
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

import numpy as np
from scipy import stats
from scipy import special as sc
from time import perf_counter
from mpmath import mp
mp.dps = 50

a, b, c, z = 1.5, 2.5, 2, 0

def _pdf_mpmath(x):
    a, b, c, z = 1.5, 2.5, 2, 0
    a, b, c, z = mp.mpf(a), mp.mpf(b), mp.mpf(c), mp.mpf(z)
    Cinv = mp.gamma(a) * mp.gamma(b) / mp.gamma(a + b) * mp.hyp2f1(c, a, a + b, -z)
    return mp.one / Cinv * x ** (a - mp.one) * (mp.one - x) ** (b - mp.one) / (mp.one + z * x) ** c

X = stats.gausshyper(a=a, b=b, c=c, z=z)
GH = stats.make_distribution(stats.gausshyper)
Y = GH(a=a, b=b, c=c, z=z)

x = np.linspace(0, 1, 1000)[1:]
tic = perf_counter()
cdf_vals = X.cdf(x)
toc = perf_counter()
t1 = toc-tic
print(f"Time for 1000 CDF value (old): {t1}")

tic = perf_counter()
cdf_vals2 = Y.cdf(x)
toc = perf_counter()
t2 = toc-tic
print(f"Time for 1000 CDF value (new): {t2}")

cdf_vals3 = np.asarray([mp.quad(_pdf_mpmath, (mp.mpf(0), mp.mpf(xi))) for xi in x], dtype=np.float64)

rerr = abs((cdf_vals - cdf_vals3)/(cdf_vals3))
rerr2 = abs((np.asarray(cdf_vals2) - cdf_vals3)/(cdf_vals3))
rerr = rerr[rerr != 0]
rerr2 = rerr2[rerr2 != 0]

print(f"Speedup: {t1/t2}")
print(f"gmean of relative error (old): {stats.gmean(rerr)}")
print(f"gmean of relative error (new): {stats.gmean(rerr2)}")

print(f"max relative error (old): {np.max(rerr)}")
print(f"max relative error (new): {np.max(rerr2)}")
```

</details>

```
Time for 1000 CDF value (main): 0.737669125024695
Time for 1000 CDF value (main): 0.006104291998781264
Speedup: 120.84433791371256
gmean of relative error (old): 8.034300408110537e-16
gmean of relative error (new): 4.553789318107634e-16
max relative error (old): 1.1235272457145405e-07
max relative error (new): 1.445515587446478e-13
```

And we get all the bells and whistles of `ContinuousDistribution`:
```python3
(Y + 1).mode()  # np.float64(1.2500000068559416)
Y.plot()
```
image

#### Additional information
Questions:
- ~~What should this be called? (Maybe see https://github.com/scipy/scipy/pull/21707#issuecomment-2412636465.)~~ We'll make it `ContinuousDistribution.from_rv_continuous`.
- ~~Should it be public? I think so, because I don't think we want to create public versions of all old distributions with the new interface.~~ Yes.

Thoughts:
- ~~Not all distributions will "just work" as written because some call other private methods of `rv_continuous` I think I can work around this.~~ This is working for almost all distributions unless they have a complicated support. We knew that the infrastructure currently required support endpoints to be either numbers or shape parameters, not functions of the shape parameters. This can be generalized in a separate PR.
- ~~Currently, the user has to pass the support of the distribution as the second argument. (If they don't, they don't get a truncated distribution... just something that is wrong sometimes.) We could eliminate this need by defining it for all instances of `rv_continuous`. (The domain of the parameters is given by `_shape_info`.)~~ 
- ~~Currently the PDF *at* support endpoints is always zero because information about whether the endpoints are included in the support is not provided explicitly by the old infrastructure. I can probably work around this or just define whether the endpoints are inclusive for each distribution.~~ Fixed by making endpoints inclusive for all distributions. The only thing this changes, I think, is that `powerlaw.pdf(0, a)` for $0 < a < 1$ is infinite (as the PDF suggests) instead of jumping to zero.
- **When we have a new discrete infrastructure, we can allow this function to accept instances of `rv_discrete`.  Similarly, we can define an interface that allows users to provide custom distribution classes. That way, they don't need to delve into the private methods of `ContinuousDistribution` to create their own distributions.**

### CODE REVIEW

**Comment by mdhaber** [comment]
Perhaps I should change this to make `ContinuousDistribution` public and expose the functionality via that name. The constructor (`__new__`) could accept an instance of `rv_continuous` and return a subclass of `ContinuousDistribution` (preferred, I think), or (if that is "wrong") the initializer could probably accept a frozen distribution and return an instance of `ContinuousDistribution`.

**Comment by fancidev** [comment]
I like the idea of making `ContinuousDistribution` public and exposing the functionality of this PR through it. Maybe something like `ContinuousDistribution.from_rv_generic(stats.gamma)` to return a (dynamic) subclass of `ContinuousDistribution`, as the PR does. This would be mirrored by `DiscreteDistribution.from_rv_discrete,` for example.

The benefit of taking an (old) distribution object rather than a frozen distribution object is that you could easily “port” the old distributions into the new framework, like you demonstrated with `Gamma = wrap_rv_continuous(gamma)`.

**Comment by mdhaber** [comment]
Yeah, I think I will give it a `from_` method. That is what we did with `Covariance` and somehow I didn't learn from it. I suppose these methods might be best in a superclass (e.g. `StatisticalDistribution.from_rv_continuous`), but then they would be inherited anyway, so I suppose it would be OK to add at the subclass level initially. I'll go ahead and do that.

File: scipy/stats/tests/test_continuous.py
Comment (mdhaber) [inline scipy/stats/tests/test_continuous.py:924]: Oops. I forgot to add a test of `sample` again.

**Comment by fancidev** [comment]
Btw, I have the impression there could be some issues related to object lifecycle etc. when a locally defined class (or anything that takes a closure) is returned. But I can’t recall the details or construct an example to demonstrate any issue.

Would it be possible, if not too clumsy, to define CustomDistribution as a regular (module level) class? That might simplify the logic somewhat, and might move closer to the idea of “offering a public interface like this for creating custom distributions so users don't need to delve into the private methods of ContinuousDistribution to create their own distributions.”

**Comment by mdhaber** [comment]
There is already a lifecycle issue with the old distributions: gh-13986. I thought about serialization issues, too, but there are workarounds if that is a problem for someone. 
There are certainly ways of doing something similar without returning a class, but there seem to be tradeoffs in terms of simplicity and/or functionality. Let's wait for a second maintainer to get involved.

**Comment by jorenham** [comment]
This is going to difficult to type 😅

File: scipy/stats/tests/test_continuous.py
Comment (steppi) [inline scipy/stats/tests/test_continuous.py:941]: Do we know enough about the situations where `make_distribution` will fail to document them? 

I'm pretty sure parameter dependent support could work in principle. You can check if the `_get_support` method in the input distribution was overridden and work from there.

Is `levy_stable` problematic because it has class variables to control the parametrization and the pdf and cdf computation methods?

For `norminvgauss`, is the problem due to a bug in the original distribution?

Not supporting circular distributions is easy to document.

For `irwinhall`, is the issue specifically because it won't work with this particular test, or is requiring an integer for a shape parameter genuinely problematic?

File: scipy/stats/_distribution_infrastructure.py
Comment (jorenham) [inline scipy/stats/_distribution_infrastructure.py:3365]: This implicitly assumes that the type of `dist` inherits directly from `rv_continuous`. Will this also work for distributions that inherit from e.g. `scipy.stats.norm_gen`?

And do you plan to support ducktyped distributions? 
Because for example there's the [`lmo.distributions.l_poly`](https://github.com/jorenham/Lmo/blob/7efe5d42ef744bda6a4edc00732a8e8938e3d493/lmo/distributions/_nonparametric.py#L73) (nonparametric (T)L-moment-based polynomial quantile-) distribution, which has a variable amount of shape parameters (L-moments actually), and therefore couldn't use `rv_continuous` as its base class. But for the most part its public methods should be compatible with that of `rv_continuous`. 

From a typing perspective it would also help to know which methods and signatures the distribution is expected to have. Because that way, I could use a structural typing approach with `typing.Protocol` to capture the wrapped distribution, and pass it as a generic type argument to the type of the returned `CustomDistribution` instance (which I'll mark it with `@type_check_only`).

Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py:3365]: > Will this also work for distributions that inherit from e.g. scipy.stats.norm_gen?

I'm not sure offhand. You're welcome to try. 

> And do you plan to support ducktyped distributions?

Yes. From the top: "Similarly, we can define an interface that allows users to provide custom distribution classes. That way, they don't need to delve into the private methods of ContinuousDistribution to create their own distributions."

File: scipy/stats/tests/test_continuous.py
Comment (mdhaber) [reply scipy/stats/tests/test_continuous.py:941]: I don't remember.

Comment (jorenham) [reply scipy/stats/tests/test_continuous.py:941]: It might be coincidence, but I noticed that `genpareto`, `genextreme`, `kappa4` and `tukeylambda` with "complicated support" are all quantile-based distribution families, in the sense that it's usually mathematically easier to work with their QFs (or the derivatives, the QDFs) than with their CDFs. I believe that `kappa4` doesn't even have a closed-form CDF. I'm not sure how this could help with this, but I thought I'd let you know either way, as it might help later on somewhere 🤷🏻

Comment (mdhaber) [reply scipy/stats/tests/test_continuous.py:941]: Those all have overrides for the `_cdf` and `_pdf`, so they should work once I override the generic method for determining the support. That is what holds them up right now ("complicated support").

(I haven't added a generic implementation of those functions in terms of the quantiles, though. So if they did not override `_pdf`, at least, that would be a problem.)

Comment (steppi) [reply scipy/stats/tests/test_continuous.py:941]: > All of those things can be improved in time. For now, I planned to document that it doesn't work with these distributions.

Yeah, I think some kind of documentation that this won't always work is the only thing needed before we can merge this. How about?

`make_distribution` is experimental and may not work with all instances of `rv_continuous`. Known failures include <list of failures>.

**Review by steppi — APPROVED** [review]
It looks like the linter added a new rule which got picked up after merging main. After the offending line is fixed, I think this is ready to merge.

**Comment by lucascolley** [comment]
> It looks like the linter added a new rule which got picked up after merging main. After the offending line is fixed, I think this is ready to merge.

It is [UP31](https://docs.astral.sh/ruff/rules/printf-string-formatting/), for which the detection in ruff isn't perfect but seems to improve frequently with releases. Probably makes sense to go through the process of disabling it (or ignoring it) until we deal with them all in one go, like https://github.com/scipy/scipy/pull/21029.

**Pull request merged**
